src/accel/detail/IntegrationSingleton.hh

```diff
@@ -0,0 +1,89 @@
+//------------------------------- -*- C++ -*- -------------------------------//
+// Copyright Celeritas contributors: see top-level COPYRIGHT file for details
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file accel/detail/IntegrationSingleton.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "accel/LocalTransporter.hh"
+#include "accel/SetupOptions.hh"
+#include "accel/SetupOptionsMessenger.hh"
+#include "accel/SharedParams.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Singletons used by the Integration interfaces.
+ *
+ * The single singleton instance contains global data. Thread-local data is
+ * managed by the \c local_transporter static class.
+ */
+class IntegrationSingleton
+{
+  public:
+    enum class Mode
+    {
+        disabled,
+        kill_offload,
+        enabled,
+        size_
+    };
+
+    // Static GLOBAL shared singleton
+    static IntegrationSingleton& instance();
+
+    // Static THREAD-LOCAL Celeritas state data
+    static LocalTransporter& local_transporter();
+
+    //// ACCESSORS ////
+
+    // Static global setup options before constructing params
+    SetupOptions& setup_options();
+
+    //! Static global setup options before or after constructing params
+    SetupOptions const& setup_options() const { return options_; }
+
+    //! Whether Celeritas is enabled
+    Mode mode() const { return mode_; }
+
+    //!@{
+    //! Static global Celeritas problem data
+    SharedParams& shared_params() { return params_; }
+    SharedParams const& shared_params() const { return params_; }
```

Comment:
Out of curiosity, what happens here if one calls `auto IntegrationSingleton::instance().shared_params()`? Will the compiler complain since you did not explicitly define its constness?

Reply:
No, `instance()` is mutable so the mutable version will be returned.